File: 0966-binary-subarrays-with-sum/0966-binary-subarrays-with-sum.cpp
int subsWithLessOrEqual(vector<int>& nums, int goal) {
    if(goal < 0) return 0;
    int n = nums.size();
    int l = 0;
    int r = 0;
    int sum = 0;
    int ans = 0;
    while(r<n) {
        sum += nums[r];
        while(l<=r && sum > goal) {
            sum -= nums[l];
            l++;
        }
        ans += (r-l+1);
        r++;
    }
    return ans;
}
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        return subsWithLessOrEqual(nums, goal) - subsWithLessOrEqual(nums, goal-1);
    }
};

File: 0966-binary-subarrays-with-sum/0966-binary-subarrays-with-sum.ts
function subsWithSumLessOrEqual(nums: number[], goal: number): number {
    const n = nums.length;
    let sum = 0;
    let r = 0;
    let l = 0;
    let count = 0;
    while(r<n) {
        sum += nums[r];
        while(l<=r && sum > goal) {
            sum -= nums[l];
            l++;
        }
        count += (r-l+1);
        r++;
    }
    return count;
}
function numSubarraysWithSum(nums: number[], goal: number): number {
    return subsWithSumLessOrEqual(nums, goal) - subsWithSumLessOrEqual(nums, goal - 1)
};